### DISCUSSION TITLE
Trying to understand connection modes and `torch.einsum()`.

### DISCUSSION THREAD

**Discussion by JSLJ23** [e3nn/e3nn#406] [Q&A] [answered]
Dear developers of e3nn,

1. Is there a part of the documentation that explains all these various connection modes, "uvw", "uvu", "uvv", etc... and how they related to the `torch.einsum()` operations that follow?
2. Also how are the Clebsch-Gordan Coefficients actually multiplied to the relevant elements of the outer product between the two input irreps of the `o3.TensorProduct`. I am trying to connect the code in `_wigner.py` but I can't see the clear link between these, some advice on this would be really appreciated.

Thank you:)

**Comment by JSLJ23** [comment]
Also I am trying to understand the activation functions within e3nn and why they require the "projection onto sphere and back" and also what actually is going on here.

1. In the example in `_s2grid.py` with the `m = ToS2Grid(6, (100, 101))`, what do the 100 and 101 res_alpha and res_beta actually mean?
2. In the next piece of example code with `x = torch.randn(3, 49)` what is this suppose to represent, are they 3 sets of irreps with a total dimension of 49 per irrep?
3. When `m(x).shape` is runned, what do the various dimensions of 3, 100, 101 mean? I see they are linked to the res_alpha and res_beta but what are the values within this output tensor in terms of their physical meaning?
4. Calling `print(m.grid.shape)` returns a shape of 100, 101, 3 which now seems to be the inverse of `m(x).shape`, is there some intuition behind this as well?

Thank you for the time to answer these questions:)

**Reply by mariogeiger** [reply, marked as answer]
1. These "uvw" etc modes are not very useful to understand actually. It's just different ways to combine the weights and the inputs. Today I [simplified a lot the code of the tensor product](https://github.com/e3nn/e3nn-jax/blob/6326cd9f9c247d778d9ddbcef89652ae5d7a1847/e3nn_jax/_src/tensor_products.py#L83-L130) in the jax implementation, it's way simpler to get familiarized with the code.
2. See [this line](https://github.com/e3nn/e3nn-jax/blob/6326cd9f9c247d778d9ddbcef89652ae5d7a1847/e3nn_jax/_src/tensor_products.py#L117)

We provide different activation functions. One of them indeed require a projection on the sphere.
1. These are the resolution for a grid on the sphere
2. You are right. `49 = 1 + 3 + 5 + 7 + 9 + 11 + 13` are the dimensions of the different irrep
3. It's a grid on the sphere along the angles alpha and beta
4. I don't remember the details

**Reply by JSLJ23** [reply]
Ok I will have a look at the Jax code, thank you for referring me to that.
With regards to the grid on sphere part, does alpha and beta correspond to the θ and φ angles respectively, or is this the other way around? Also if one of them represents φ from 0 to 2π, shouldn't the resolution of that be double for equally interpolated points on the grid?

**Reply by mariogeiger** [reply]
beta is the angle going from 0 to pi and alpha goes from 0 to 2pi. 

Yes I agree, I would most of the time double the resolution along alpha.

**Reply by JSLJ23** [reply]
Got it, thank you:)
